File: python/productExceptSelf.py
# https://leetcode.com/problems/product-of-array-except-self/

from typing import List


class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        """
        Return an array where each element is the product of all others.

        For example [1,2,3,4] ->

        Multiply answer[i] by prefix product, before nums[i] from the left.

        prefix @ nums[0] -> $1
        prefix @ nums[1] -> $1 * 1
        prefix @ nums[2] -> $1 * 1 * 2
        prefix @ nums[3] -> $1 * 1 * 2 * 3

        Multiply answer[i] by suffix product, before nums[i] from the right.

        suffix @ nums[3] -> $1
        suffix @ nums[2] -> $1 * 4
        suffix @ nums[1] -> $1 * 4 * 3
        suffix @ nums[0] -> $1 * 4 * 3 * 2

        Complexity:
        - Time: O(n)
        - Space: O(1)
        """
        n = len(nums)
        answer = [1] * n
        left_prefix, right_suffix = 1, 1
        for left_idx in range(n):
            right_idx = n - left_idx - 1
            left_num, right_num = nums[left_idx], nums[right_idx]

            # Update answer with prefix and suffix products
            answer[left_idx] *= left_prefix
            answer[right_idx] *= right_suffix

            # Update prefix and suffix products
            left_prefix *= left_num
            right_suffix *= right_num
        return answer

File: java/ProductExceptSelf.java
// https://leetcode.com/problems/product-of-array-except-self/

import java.util.Arrays;

public class ProductExceptSelf {
    /**
     * Product of all the elements of {@code nums} except
     * {@code nums[i]}. What we need to do is:
     *
     * <ul>
     *     <li>Use cumulative product of left side without l-1</li>
     *     <li>Use cumulative product of right side without r+1</li>
     *     <li>Left product fills from left to right</li>
     *     <li>Right product fills from right to left</li>
     * </ul>
     *
     * Here is an example for demonstrative purposes:
     *
     * <pre>
     * [1, 2, 3, 4] (original)
     * [1, 1, 2, 6] (left transform)
     * [24, 12, 4, 1] (right transform)
     * [24, 12, 8, 6] (combined transform)
     * </pre>
     *
     * <p>Complexity:
     * <ul>
     *     <li>Time: O(n)</li>
     *     <li>Space: O(1)</li>
     * </ul>
     */
    public int[] productExceptSelf(int[] nums) {
        int leftProduct = 1;
        int rightProduct = 1;

        int[] result = new int[nums.length];
        Arrays.fill(result, 1);

        // Apply left and right transforms together
        for (int l = 0, r = result.length - 1; l < result.length; l++, r--) {
            result[l] *= leftProduct;
            leftProduct *= nums[l];

            result[r] *= rightProduct;
            rightProduct *= nums[r];
        }

        return result;
    }
}
